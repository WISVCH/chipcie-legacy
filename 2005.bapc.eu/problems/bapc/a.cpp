/*
  [NKP'05] ASM
  by Jan Kuipers
*/

using namespace std;

#include <iostream>
#include <vector>
#include <string>
#include <queue>

const int dy[4] = { 1,-1, 0, 0 };
const int dx[4] = { 0, 0, 1,-1 };

int main () {

  int runs;
  cin >> runs;

  while (runs--) {
    int Y,X,N,H;
    cin >>Y>>X>>N>>H;
    
    vector<string> s(Y);
    for (int y=0; y<Y; y++) cin >>s[y];

    vector<vector<int> > t(Y, vector<int>(X, 0));
    for (int y=0; y<Y; y++)
      for (int x=0; x<X; x++)
        t[y][x] = s[y][x]-'0';
    
    for (int i=0; i<N; i++) {
      int y,x;
      cin >>y>>x;
      t[y-1][x-1]++;
    }

    bool again = true;
    while (again) {
      again = false;
      for (int y=0; y<Y; y++)
        for (int x=0; x<X; x++)
          if (t[y][x] > H) {
    	    t[y][x]-=4;
            again=true;
 	  
  	    for (int d=0; d<4; d++) {
	      int ny=y+dy[d], nx=x+dx[d];
  	      if (nx<0 || nx>=X || ny<0 || ny>=Y) continue;
	      t[ny][nx]++;
	    }
          }
    }

    for (int y=0; y<Y; y++) {
      for (int x=0; x<X; x++)
        cout << char('0'+t[y][x]);
      cout << endl;
    }
  }

  return 0;
}

  
