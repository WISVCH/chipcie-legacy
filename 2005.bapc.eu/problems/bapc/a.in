11
3 4 1 5
1023
2344
0000
2 3
3 4 1 4
1023
2344
0000
2 3
7 7 1 9
9999999
9999999
9999999
9999999
9999999
9999999
9999999
4 4
7 7 10 9
9999999
9999999
9999999
9999999
9999999
9999999
9999999
4 4
4 4
4 4
4 4
4 4
4 4
4 4
4 4
4 4
4 4
1 1 0 7
7
1 1 1 7
7
1 1
1 1 3 3
0
1 1
1 1
1 1
1 1 4 3
0
1 1
1 1
1 1
1 1
100 100 100 9
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
9999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
50 50
100 100 100 4
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
11 38
68 57
63 13
30 37
3 47
53 57
100 54
46 48
8 6
88 95
66 92
54 93
18 69
42 3
10 10
48 24
81 72
37 67
55 95
41 66
74 57
11 46
93 25
30 8
55 52
43 69
61 20
88 96
51 30
92 29
5 12
36 91
57 85
79 72
89 12
20 72
17 62
86 82
90 61
12 67
35 44
15 54
1 47
76 2
82 51
63 19
9 86
70 50
21 18
29 1
72 61
22 100
82 40
100 59
25 23
18 11
64 59
5 32
85 64
67 32
51 66
4 81
30 59
76 25
25 50
62 56
8 48
87 83
93 89
63 86
96 17
27 32
15 11
75 31
62 99
16 93
73 12
23 19
96 2
3 50
6 20
68 64
98 65
5 54
92 90
6 67
50 39
50 32
14 64
14 76
68 27
91 29
100 40
93 13
14 47
66 95
10 71
35 85
39 7
96 39
100 100 100 7
7613172415232336424030756263371262571032555701643256344162411547016111476167243670424033274440342263
3222311654445061114301751313577011252330074573604345432563133044161424423073353007543721235531247301
0433526002001554476723454613225260533443643720467251526100422140224560044336475312564375347554676244
2406035432744431026365141737103343147602006541644402116302212646153043243110274634155705126302220524
0504624010755022720146451671315465047451446140432446213372434172777634770514557714635174437056245120
5205711561475222416056024262402122723071130663024721523141302624016415531617121513765271432614215156
6317136257762647131563377514662514427344242526431500037710377640213147563736501652665656616572613127
0764612313175053621437222035155143524756375072354510432636443160432030667366523274230423260676631034
0227515330525755370650446270234357220053025512252337374213236663516513016577014247106527143721302153
4652342356526724444507520023106563017255027611365735607602112070371316011702030524204165306505534765
5676571521245442627125723467345134703065201755137042634102037542142440161266771705161571074751165331
4405763654551632423420371456707357045323700164326760717055655502516241541657202070162067544325576412
5766535747730022101656173761373062635321325327537624433721051457633375527252125007553045253310274523
3752124242341014062235545031554051363717144345743176642445523220356757663217014323217454226540506373
3516036042365574240476137635120401304100337016433572305261773743173503637530373642002620727213622102
5654304607541543373212526350453130561141015172421752224051016331407524625344507604026523624567620672
2547004613417535250074554474266432343724265130655664332771310153317601230743731417142332463471032712
1451324355764037361054445477412557601236035430376473130607640015776111714460417366671651635636435247
3300060407765656420265114541700241247207105662525553264631021045171112032610062537065540642757566677
0022642625464442102755724402201304560042500154365662354714134264642647117124663341576167572211605161
0727144727260756035527740645726720635227146133737401405461254046121653567302775645705543660274007164
4323725622460365021711106205361013306160326364166257377117645747307116140706315132161002067563413424
2312211526650732034023507212637060201353130233437031401322505403433560003026461364425557730470743212
2225244625102537206136537472604026652234755120050764547406776371256471176601616604250121710647765534
6435246054650121242252021706433276712517104217333551753043707236020301211732666133221535126041643673
0251104772025356004122163026622647057446641376206610036330112306203247024153773555550303424654504020
0224407373150620233657424120241273633563070062005362325747764003466734244242552200532226124524062666
3027361033243775124340065157735237166261515005627253251766761405423242323123605522050146034105647263
4157202000532502741705002741125532462663646116302472202476600354512070364176726176416655535562143342
4700076725013421376033515750212603702574205547677472753441462753243413143617607546743207356643167420
7342262061527463672240003162503547765163236274553777077305666132020646161700455755754605463276400464
2733733310265132037422610210675070363614632455651513774716775474422003467724401552041532331070542742
3002736437111252714245235777731241434270111236427043567266152276331051062105747743211137044263016113
2114211250113325554511774713026340761107110553220672761452755501507710621676110207475532020500650541
5247135256455043605625326106640474307024671476756430263073667626256251640007675430457756244226053771
0550557344775454322667015061752727704730337074427705606314401604574167122121574464245076546626206614
6360412106662027015554402600515343304524302604506363723103255701242714354634204032424745362053177360
7253007214346773630112165354645662163104530327612623141677353031257567133265156431745134061365402360
2735113361722667713775252664144437150407671057051340076254670233140210177275172236335163542575001122
1311516700136474557211706017132273344135150017474316416217123342706312037130077521362103115451641473
6065116105626100303410062623226124635445237047177431537712244156531305020025734700053745161576351401
2032067014714767624015503155556762260161553141134535305613370266405423571015220673324015344463223765
3454461602475721037373766202177434072262422124327256554475604557256474737151503420275734415072717557
2421572275615113047464456366035207357506260071375233770526332052017767516515645460750032665073274262
1337145702372342375162255441624057764464613140377015332076260666554122604110150151604113731017665207
5671711161232046152103113770177656757165201505322532143432442337124132662733465640054317653006711424
6121054541301455764537536547173706003457517654342725770545454754554122073060414716406552210605262674
0743723430346744417627434616503577171434461355727601554135005365274337705032624515662305006634353706
6763015642357131436534360633532330632316543456522475024107453606761112776233015254056466333613401513
7025251366533511170523347040563565213341215576605650214110166535266512733402273750771302421274712574
7673205470341342645266552475261140460227254300674313160337055161237251277637616240556601516325440475
5144774502742511322031356026413127617267454725057262310112753275256174031424221157401422610147664473
0761415637706317732542601465050056217773663524417674054123620635556445734306747736334756230323070735
4300606546474730706137452543400030110764534227217132070244504517520517272614536441640675345726770752
6710525253624464351627545436400271343060342711346520474140700720054353407670734606646352352344352007
3402373227026764246510054426436776125450452241665436443006741530343010055372450114750250044117243554
5512004455561431711755174337004565766433102255346434143576477045643406717134760523130407347253737277
1600034714437667267536701631120201204676241120146021611704016137257637453665671473655755456261106171
0363140741435212167535026124215253567563727440656531233046477114473341733677654427542147706617356613
7062661435065527676574414762575474652715413176155734475064233073450642303722530223662460013312657733
1634557207432314704717221170023106564404357501171562404513515025703344701751070146316760241744434060
4605566556614122012254110554534115663430211624125342753730304025504057071163576222420013053455226623
5226302170322042055021576714537252112322354550055560637521174112323555702360365133126070107611043321
0112572050113232322422243206527541010216237652004256472723675551652774527404605222071363643320505774
3462067664016107463236556253257516024703314124062305152700225276070461712324732262707277221741540607
7001335367541440640061722563362262325053720064050056250426755173425122514513110225042104701410052574
0454167207425567664565206250745702411341303051737316136753740440702036166464567410223010415451441644
4533117676317532653260631202146521033722653527704222205335542414527011376641541164314147623065437440
5704545306472007146762047406046636203733502002717505702652456331151445025225516466167344511345452627
3210435143521700254066323075614135100614177267304702544044625231041022531457300000254760342065161273
5466046346346623503046023215400147043677434216566162764205755061765245100531307115303034521327416232
7537260667704307333064304022254127055034724354200616246640066570475270663210641622446333331101250770
0563703544466025350711012263253221511075533336073064266414732254426525370633432040467220723140502364
0140774426467046376211351063447453542105743650407220355453717764231541236523563415442217501577122276
4112635420035577706417106320516120313055050520410361272124106010051356164136100236355560200001006234
0424653753101566246446440546702744422727230306132077643610500774437621544575310617573064034024567442
5261766376106671656702736416060506774523433213300171464235033013703355711232572601740373076407777225
0161423116717667152241640530065073135456240236143467544527525225630375222245051310264646610342025155
6770135107427703542362725742017716721542404374651007371075276267062236476026403513443542261100016231
0707725320133576330522622730161251143605611160014076250444462500614174755064065474212366724104165604
2710704761456670156073277365161453443031177766773602122050576733277572602500377657161377346131450327
5177677277046130427765026476725534414433347152124013525364166631373736262100412023375034542335564050
6761061403526014140612147623673562660730202004410471615577166434613706726226670637020570506351032322
1247465255015035625333604636160204143760461174757402062455363313027315364573160030236473222563165017
6452145335365613207431214371300145362015543224654517630267327434172373057372757420304422364320737176
4733335204635231663140460317405001434154647363441755723754625425560105561517053141643130522265733743
5116626671742505636450635661524246120710004364047605770457521255060051051407014074563530403620326333
4306015021016501332735565103436176777471603453506701457160433242131071251516077667734743716232545654
7720437426215141004712254711566550513456277713113515032423672447421766516407710371074236516752714013
6644343264555542770516461305310274621257624300577640417547300327002127001441411375446313243205105310
2104545166454012335570775173403713360076743345500357357166427710247256042477147174031747602573510445
3415004143240371231033526336773244044461005034167262544477267613217755056551177016272631557430761757
4542224313321513344024554436305051273624257320755367734470230736055440065710005543436076011104711455
4532552557213651647760064776343027274441363640024712011400245647566122250544546230431502626302250062
2172636474634266060616110370352561755514177616414475207236263432527270717670432171612535354717162011
0027617341433245072346256340351354235262325652351606433267224451005523756543770050623242174631731403
0016551441421345507673424456061062637774036262732622164522330536721710323142335517726470633607471562
5140202530347265666411210601374000131304331367045616037020257667672723466514411202005007025703672164
4222737450652657657604507471565100303240225402470350730670746456406032054625064011004163157714565460
6653470554565662745015731416076754223271647326512314007150057644266605762624476727320245242127555436
11 38
68 57
63 13
30 37
3 47
53 57
100 54
46 48
8 6
88 95
66 92
54 93
18 69
42 3
10 10
48 24
81 72
37 67
55 95
41 66
74 57
11 46
93 25
30 8
55 52
43 69
61 20
88 96
51 30
92 29
5 12
36 91
57 85
79 72
89 12
20 72
17 62
86 82
90 61
12 67
35 44
15 54
1 47
76 2
82 51
63 19
9 86
70 50
21 18
29 1
72 61
22 100
82 40
100 59
25 23
18 11
64 59
5 32
85 64
67 32
51 66
4 81
30 59
76 25
25 50
62 56
8 48
87 83
93 89
63 86
96 17
27 32
15 11
75 31
62 99
16 93
73 12
23 19
96 2
3 50
6 20
68 64
98 65
5 54
92 90
6 67
50 39
50 32
14 64
14 76
68 27
91 29
100 40
93 13
14 47
66 95
10 71
35 85
39 7
96 39
