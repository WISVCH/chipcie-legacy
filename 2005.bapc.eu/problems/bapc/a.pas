program asm_bram;

{$mode objfpc}{$H+}

uses
  Classes, SysUtils;
  
const
  MIN_X = 1; MAX_X = 100;
  MIN_Y = 1; MAX_Y = 100;
  MIN_N = 1; MAX_N = 100;
  MIN_H = 3; MAX_H = 9;

var
  numTestCases,testCase:integer;
  x,y,n,h:integer;
  lattice,dropped:array[MIN_X..MAX_X, MIN_Y..MAX_Y] of integer;
  i,j:integer;
  c:char;
  s:string;
  xi,yi:integer;
  toppleCount,totalTopples:integer;

begin

  readln(numTestCases);
  for testCase:=1 to numTestCases do begin
    try
      // read & check input
      readln(y,x,n,h);
      if (x<MIN_X) or (x>MAX_X) then Exception.CreateFmt('x = %d is not the range %d..%d',[x,MIN_X,MAX_X]);
      if (y<MIN_Y) or (y>MAX_Y) then Exception.CreateFmt('y = %d is not the range %d..%d',[y,MIN_Y,MAX_Y]);
      if (n<MIN_N) or (n>MAX_N) then Exception.CreateFmt('n = %d is not the range %d..%d',[n,MIN_N,MAX_N]);
      if (h<MIN_H) or (h>MAX_H) then Exception.CreateFmt('h = %d is not the range %d..%d',[h,MIN_H,MAX_H]);
      for j:=1 to y do begin
        for i:=1 to x do begin
          read(c);
          if not (c in ['0'..'9']) then raise Exception.CreateFmt('Character %c in lattice discription, line %d, column %d (should be in the range ''0''..''9'')',[c,j,i]);
          lattice[i,j]:=Ord(c)-Ord('0');
          if lattice[i,j]>h then raise Exception.CreateFmt('Pile exceeds critical height: %d > %d, line %d, column %d',[lattice[i,j],h,j,i]);
          dropped[i,j]:=0; // inits dropped array
        end;
        readln(s);
        if s<>'' then raise Exception.CreateFmt('Trailing characters at end of lattice description, line %d: %s',[j,s]);
      end;
      for i:=1 to n do begin
        readln(yi,xi);
        if (xi<MIN_X) or (xi>x) then raise Exception.CreateFmt('x position of grain %d is %d, which is outside of the range %d..%d',[i,xi,MIN_X,x]);
        if (yi<MIN_Y) or (yi>y) then raise Exception.CreateFmt('y position of grain %d is %d, which is outside of the range %d..%d',[i,yi,MIN_Y,y]);
        Inc(dropped[xi,yi]);
      end;
      // run iterations
      repeat
        totalTopples:=0;
        //  add dropped to lattice, clear dropped
        for i:=1 to x do for j:=1 to y do begin
          Inc(lattice[i,j],dropped[i,j]);
          dropped[i,j]:=0;
        end;
        //  topple grains from lattice to dropped
        for i:=1 to x do for j:=1 to y do begin
          if lattice[i,j]>h then begin
            toppleCount:=(lattice[i,j]-h+3) div 4;
            if i>MIN_X then Inc(dropped[i-1,j],toppleCount);
            if i<x     then Inc(dropped[i+1,j],toppleCount);
            if j>MIN_Y then Inc(dropped[i,j-1],toppleCount);
            if j<y     then Inc(dropped[i,j+1],toppleCount);
            Dec(lattice[i,j],4*toppleCount);
            Inc(totalTopples,toppleCount);
          end;
        end;
      until totalTopples=0;
      // write solution
      for j:=1 to y do begin
        for i:=1 to x do
          write(Chr(Ord('0')+lattice[i,j]));
        writeln;
      end;
    except
      on E:Exception do begin
        raise Exception.CreateFmt('Error in test case %d: %s',[testCase, E.message]);
      end;
    end;
  end;
  if not Eof then raise Exception.CreateFmt('Not at end of file at end of testcases',[]);

end.

