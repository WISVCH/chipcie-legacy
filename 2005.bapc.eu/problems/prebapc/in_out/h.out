0
71
104
2676
3763
895402
81267
50541
