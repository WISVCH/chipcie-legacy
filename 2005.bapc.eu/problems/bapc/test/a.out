94
157
0
0
261799416
